src/providers/sn_provider/starknet_provider.rs

```diff
@@ -0,0 +1,63 @@
+use crate::{
+    into_via_wrapper,
+    models::felt::Felt252Wrapper,
+    providers::eth_provider::{
+        error::ExecutionError,
+        starknet::{ERC20Reader, STARKNET_NATIVE_TOKEN},
+        utils::{class_hash_not_declared, contract_not_found},
+    },
+};
+use reth_primitives::U256;
+use starknet::core::types::{BlockId, Felt};
+use std::sync::Arc;
+use tracing::Instrument;
+
+/// A provider wrapper around the Starknet provider to expose utility methods.
+#[derive(Debug, Clone)]
+pub struct StarknetProvider<SP: starknet::providers::Provider + Send + Sync> {
+    /// The underlying Starknet provider wrapped in an [`Arc`] for shared ownership across threads.
+    provider: Arc<SP>,
+}
```

Comment:
Here I wonder if we can't go even further and replace:

```rust
pub struct EthDataProvider<SP: starknet::providers::Provider + Send + Sync> {
    database: Database,
    starknet_provider: SP,
    pub(crate) chain_id: u64,
}
```

by something like:

```rust
pub struct EthDataProvider<SP: starknet::providers::Provider + Send + Sync> {
    database: Database,
    starknet_provider: StarknetProvider<SP>,
    pub(crate) chain_id: u64,
}
```

Maybe we could address this in a follow up PR, what do you think?

Reply:
yeah let's do it in another PR